### ISSUE TITLE
Add multiple phone numbers to person entity (case, contact, event ) [5]

### ISSUE THREAD

**Issue by bernardsilenou** [hzi-braunschweig/SORMAS-Project#2744]
### Proposed Change
- Add 2 more fields to capture other phone numbers
- Show these fields only after selecting "multiple phone numbers "
- Migration: use phone number 1 as formal phone number variable. If more than one numbers were entered in this field, then users have to manually transfer them to phone number 2
- First on web and mobile is optional
![image](https://user-images.githubusercontent.com/12996724/91655902-93a07200-eab4-11ea-9c20-927df7d2e638.png)

**Comment by markusmann-vg** [comment]
- Add also Selection (Festnetz privat, Mobil, Arbeit, others) => others lead to free text

**Comment by FredrikSchaeferVitagroup** [comment]
Implementing several phone numbers will probably have some influence on #2638 .

**Comment by kwa20** [comment]
Since the primary phone number field is connected to other entities (e.g. adoption of number in creation forms when converting contacts to cases, referenced number for sms messages), what do you think about the approach to treat this similarly to what we have for addresses?

<img width="822" alt="Screen Shot 2020-11-04 at 1 21 29 PM" src="https://user-images.githubusercontent.com/70317594/98111316-b2e6c180-1ea0-11eb-9267-3ce2cb9c0e86.png">

**Comment by ChristopherRiedel** [comment]
@kwa20 I get your approach, but at least for the addresses the current status is only temporary. (See #2869)
Maybe we should handle phone numbers in a similar way.

**Comment by kwa20** [comment]
@ChristopherRiedel I agree, I think continuity is important here, also when considering user friendliness. This also doesn't seem to be something that is implemented as a first step, second step but rather a change of a change if I understand correctly. We should then implement it consistently for addresses and phone numbers right away.

**Comment by bernardsilenou** [comment]
@ChristopherRiedel  we though of having a table on the db for contact information, just as we have for addresses. The reason is that there may be multiple contacts and email addressed and each of them liked to the person, family member (father , mother, uncle), legal organization (eg fro homeless people), etc. Users need to know who each of the multiple contacts are linked to
what do you think about this?
@kwa20  would add a mockup

**Comment by kwa20** [comment]
@ChristopherRiedel This is what we've come up with as a possible solution:

<img width="923" alt="Screen Shot 2020-11-09 at 8 07 08 PM" src="https://user-images.githubusercontent.com/70317594/98585921-4157aa80-22c8-11eb-8356-b40aa24c343b.png">

If I understand correctly (reading #2869),  you want to get rid of the separation of fields and the address list. The advantage that the separate fields have though, is that they're clearly associated to the person that the form is about, while additional contact information entries might be associated to people related to that person. Users then don't have to navigate through "new entry > done > save" for initial entries. The fields are the primary contact information by default, as long as no other entries in the additional list have been selected as primary.

'Owner of phone' would be migrated to the person type. It would not make much sense in the contact information anymore. I suggest that if 'owner of phone' was filled previously, entries should appear in the additional list with set to primary. 
@bernardsilenou  do you agree?
@Candice-Louw If necessary, we might also add the 'phone number type' to the initial contact information.

**Comment by bernardsilenou** [comment]
@kwa20 Nice design,  minor comments:
1. Since phone number type is needed, then we need to add it to the "contact information" section also. Since the type is a subset of the phone number, I would put phone number type beside phone number and only show it when a phone number is registered. We can have a situation with only email contact, thus no need to show phone number type
2. What do you think about replace person type with something like "contact owner" because the legal guardian/ Other can be an organization

3. Do not show first and last name when person type/ contact owner == "case"  or contact or event participant. We matches the option here according to the entity. Case person if the person is viewed through the case from and contact likewise. Tricky for a person linked to both contact and case.
4. If only contact information is provided, set it as primary contact. users can then override it if there are many others

**Comment by FredrikSchaeferVitagroup** [comment]
@HolgerReiseVSys 
This will probably influence the SurvNet interface.

**Comment by kwa20** [comment]
@bernardsilenou I guess your 2nd point makes sense. However, we would then be able to deposit contact information of facilities. It would maybe then make sense to also be able to chose such a facility: Contact owner (person type) > facility > facility drop down appears > chosen facility shown as contact owner (person) in list (e.g. orphanage). It might be difficult to do it in this issue but could be a second step option. Such contact info could run under 'other' initially.

**Comment by BarnaBartha** [comment]
@markusmann-vg This needs to be clarified/refiend on what we wish to achieve here + re-estimated.

**Comment by MateStrysewske** [comment]
@bernardsilenou @kwa20 @markusmann-vg Based on your refinements and comments so far, I propose the following solution:

* Introduce a new `PersonContactDetails` entity with the following fields:
    * `person`: The person associated with these contact details
    * `thirdParty` (boolean): True if the contact details belong to someone other than the person in question
    * `thirdPartyRole` (text): Free text field, e.g. mother, father, legal guardian, general practitioner, community leader, retirement home, ...
    * `thirdPartyName` (text): Name of the third party; I don't think we need to split this into first and last name
    * `type` (enum: PersonContactDetailsType): Phone/Email/Other/*TBD - is there anything else we need to consider?*
    * `phoneNumberType` (enum): Landline, Mobile, Work, Other
    * `details` (text): Specification of the "Other" person contact details type or the "Other" phone number type
    * `contactInformation` (text): The actual information, e.g. phone number or email address
    * `additionalInformation` (text)
    * `primary` (boolean): Whether these are the primary contact details of the person
* The entity basically works the same as exposures for epi data, i.e. it's an embedded entity that is synchronized alongside the person

I would suggest to strictly map one PersonContactDetails instance to one actual piece of contact information, e.g. don't have one instance that contains both a phone number and email address (like in the screenshot above) because that makes the system more intuitive. For example, when you add two phone numbers for the same person (landline and mobile), where do you put the email address?

I also don't think we necessarily have to define the role of the third party. There are A LOT of possible options here, and there's no way we could think of a complete list. I'd make it a free text field instead since I don't think it's very likely that we will have to filter by its value at some point in the future.

`Primary` should probably not be totally exclusive, but there should be an option to specify both a primary phone number and a primary email address.

* Add a list of PersonContactDetails to the person form, similar to the list of addresses, with the following columns:
    * Edit icon (no caption)
    * Primary (displays a checkmark if true, empty otherwise)
    * Owner (either "This person" if it's the referenced person or the third party role)
    * Owner Name (the person's name or the third party name)
    * Contact Type (=> person contact details type; including the phone number type or the details field in brackets, if available)
    * Contact Information
* Add an edit form that opens in a popup window
    * First row should be a checkbox mapped to `thirdParty` with the caption "Collect contact details of another person or facility"
    * `thirdPartyRole` and `thirdPartyName` are only visible when `thirdParty` is checked
    * `phoneNumberType` is only visible for `type` "Phone"
    * `details` is only visible for `type` "Other" or `phoneNumberType` "Other"
    * `primary` should also be a checkbox, last row in the form, caption "Primary contact details"
    * If `primary` is enabled, saving the form should check whether there is another primary contact details entity for the same person and the same PersonContactDetailsType; if that's the case, a confirmation popup with the following text is shown: "There already are primary contact details of this type recorded for this person. Do you want to set these contact details as the primary contact details instead?"
    * Validation of the `contactInformation` field depends on the selected `type`
 * Add a new user right `PERSON_CONTACT_DETAILS_DELETE`, for admins, admin supervisors and national users
 * Write an SQL migration script that creates new entities for the contact details of existing persons (@bernardsilenou @kwa20 Please check if this is okay!)
     * Add a new entry when `phone` is filled in; if `phoneOwner` is also filled in, set `thirdParty` to true and map `phoneOwner` to `thirdPartyRole` AND `thirdPartyName`
     * Add a new entry when `emailAddress` is filled in
     * Add a new entry when `generalPractitionerDetails` is filled in; `thirdParty` is true, `thirdPartyRole` is "General practitioner" and `thirdPartyName` maps to `generalPractitionerDetails`
 * Remove `phone`, `phoneOwner`, `emailAddress` and `generalPractitionerDetails` from the Person entity
 * Check any code that uses these fields (e.g. search fields) and re-write it to use the person's PersonContactDetails instead
 * Keep the field names "phone" and "emailAddress" in CaseExportDto and ContactExportDto and rename the i18n person properties to "Primary phone number" and "Primary email address"; we don't have to write a migration script for export configurations this way; `phone` should therefore contain the primary phone number (only the number) and `emailAddress` the primary email address. Add another column `otherContactDetails` (includes the `contactInformation` value plus the `type` value in brackets; `details` instead if `type` is "Other");

**Comment by bernardsilenou** [comment]
@MateStrysewske @kwa20 
The main comment I have is of we can map both email and tel as attributes to  `PersonContactDetails` with as least one provided as required to create a new entity. I commented below but  I think your the once to decide.

> @bernardsilenou @kwa20 @markusmann-vg Based on your refinements and comments so far, I propose the following solution:
> 
> * Introduce a new `PersonContactDetails` entity with the following fields:
>   
>   * `person`: The person associated with these contact details
>   * `thirdParty` (boolean): True if the contact details belong to someone other than the person in question
>   * `thirdPartyRole` (text): Free text field, e.g. mother, father, legal guardian, general practitioner, community leader, retirement home, ... 
OK, I do not think we need filter for this

>   * `thirdPartyName` (text): Name of the third party; I don't think we need to split this into first and last name

I agree with you also.

>   * `type` (enum: PersonContactDetailsType): Phone/Email/Other/_TBD - is there anything else we need to consider?_

I wrote this before reading your comment:
Contact details always has a phone  + email (one of them being requited), do we then need this attribute? 
Your specification is the optimum scenario. To me, I would suggest that we keep both together, if users set a primary contact, then both email and phone are used as primary. this also reduces the work of the users, not to create 2 records, one for phone and another for email. ie we force the users to always use correct email and tel on the primary record form.

>   * `phoneNumberType` (enum): Landline, Mobile, Work, Other

@kwa20 users can have a mobile phone for work purpose and a landline at home. I am not sure if the goal of this variable is to capture the type of tel (landline or mobile) or the purpose of the tel (work, personal). @kwa20 can you check with SurvNet?
If not clearly specified in SurvNet, then it would be better to user:
 `phoneNumberType`(enum): Landline, Mobile, Unknown  and also add
 `phoneNumberPurpose`(enum):Personal, Work, Unknown, Others + free txt

>   * `details` (text): Specification of the "Other" person contact details type or the "Other" phone number type
>   * `contactInformation` (text): The actual information, e.g. phone number or email address
>   * `additionalInformation` (text)
>   * `primary` (boolean): Whether these are the primary contact details of the person
> * The entity basically works the same as exposures for epi data, i.e. it's an embedded entity that is synchronized alongside the person
> 
> I would suggest to strictly map one PersonContactDetails instance to one actual piece of contact information, e.g. don't have one instance that contains both a phone number and email address (like in the screenshot above) because that makes the system more intuitive. For example, when you add two phone numbers for the same person (landline and mobile), where do you put the email address?
> 
> I also don't think we necessarily have to define the role of the third party. There are A LOT of possible options here, and there's no way we could think of a complete list. I'd make it a free text field instead since I don't think it's very likely that we will have to filter by its value at some point in the future.

I agree with you also. 
> 
> `Primary` should probably not be totally exclusive, but there should be an option to specify both a primary phone number and a primary email address.
> 
> * Add a list of PersonContactDetails to the person form, similar to the list of addresses, with the following columns:
>   
>   * Edit icon (no caption)
>   * Primary (displays a checkmark if true, empty otherwise)
>   * Owner (either "This person" if it's the referenced person or the third party role)
>   * Owner Name (the person's name or the third party name)

This means we show both names of the person in one column since third party name is one variable.

>   * Contact Type (=> person contact details type; including the phone number type or the details field in brackets, if available)
>   * Contact Information
> * Add an edit form that opens in a popup window
>   
>   * First row should be a checkbox mapped to `thirdParty` with the caption "Collect contact details of another person or facility"
>   * `thirdPartyRole` and `thirdPartyName` are only visible when `thirdParty` is checked
>   * `phoneNumberType` is only visible for `type` "Phone"
>   * `details` is only visible for `type` "Other" or `phoneNumberType` "Other"
>   * `primary` should also be a checkbox, last row in the form, caption "Primary contact details"
>   * If `primary` is enabled, saving the form should check whether there is another primary contact details entity for the same person and the same PersonContactDetailsType; if that's the case, a confirmation popup with the following text is shown: "There already are primary contact details of this type recorded for this person. Do you want to set these contact details as the primary contact details instead?"
>   * Validation of the `contactInformation` field depends on the selected `type`
> * Add a new user right `PERSON_CONTACT_DETAILS_DELETE`, for admins, admin supervisors and national users
> * Write an SQL migration script that creates new entities for the contact details of existing persons (@bernardsilenou @kwa20 Please check if this is okay!)
>   
>   * Add a new entry when `phone` is filled in; if `phoneOwner` is also filled in, set `thirdParty` to true and map `phoneOwner` to `thirdPartyRole` AND `thirdPartyName`

the phone owner can be the same name as the person (second phone number of the same case person), thus may be nice to additional check it the name is "exact match". If yes, map to `phoneOwner`. If this may be challenging in the SQL, then we can drop it, it cause no harm.
>   * Add a new entry when `emailAddress` is filled in
>   * Add a new entry when `generalPractitionerDetails` is filled in; `thirdParty` is true, `thirdPartyRole` is "General practitioner" and `thirdPartyName` maps to `generalPractitionerDetails`
> * Remove `phone`, `phoneOwner`, `emailAddress` and `generalPractitionerDetails` from the Person entity
> * Check any code that uses these fields (e.g. search fields) and re-write it to use the person's PersonContactDetails instead
> * Keep the field names "phone" and "emailAddress" in CaseExportDto and ContactExportDto and rename the i18n person properties to "Primary phone number" and "Primary email address"; we don't have to write a migration script for export configurations this way; `phone` should therefore contain the primary phone number (only the number) and `emailAddress` the primary email address. Add another column `otherContactDetails` (includes the `contactInformation` value plus the `type` value in brackets; `details` instead if `type` is "Other");

**Comment by kwa20** [comment]
@MateStrysewske Nice summary! 

> Check any code that uses these fields (e.g. search fields) and re-write it to use the person's PersonContactDetails instead

This would also apply to the symptom journal interfaces. These should always use the primary PersonContactDetails. Changing the primary details would also mean that interfaces would have to be synced.

> @kwa20 users can have a mobile phone for work purpose and a landline at home. I am not sure if the goal of this variable is to capture the type of tel (landline or mobile) or the purpose of the tel (work, personal). @kwa20 can you check with SurvNet?
If not clearly specified in SurvNet, then it would be better to user:
phoneNumberType(enum): Landline, Mobile, Unknown and also add
phoneNumberPurpose(enum):Personal, Work, Unknown, Others + free txt

@bernardsilenou SurvNet does not differentiate this. Your approach is also fine with me.

**Comment by Candice-Louw** [comment]
@HolgerReiseVSys - please double-check for SurvNet compatibility/mapping.

**Comment by Candice-Louw** [comment]
CH MED perspective:

var | Title de-CH | Description de-CH | Format 
-- | -- | -- | --
phone_number (Fall) | Festnetznummer | Festnetznummer des COVID-19-Falls | CharacterString(+xxxxxxxxxxx)
mobile_number (Fall) | Mobil | Mobiltelefonnummer des COVID-19-Falls | CharacterString(+xxxxxxxxxxx)
phone_number (Kontakt) | Festnetz | Festnetznummer der Kontaktperson | CharacterString(+xxxxxxxxxxx)
mobile_number (Kontakt) | Mobil | Mobiltelefonnummer der Kontaktperson | CharacterString(+xxxxxxxxxxx)

**Comment by VitaliHZI** [comment]
Hello Team, 

i tested it on https://test.sormas.netzlink.com/sormas-ui/ and if click the checkbox "primary contact details" my entry is not saved and not displayed. As I learned in a test stage no bugs should be opened, i think this tickets should be reopened and need to be fixed. 


![image](https://user-images.githubusercontent.com/79860634/111755235-307ffc80-8899-11eb-8dda-8a0a4d36fe00.png)



BR 
Vitali

**Comment by MateStrysewske** [comment]
@VitaliHZI Thanks! Looks like there's something not working correctly there indeed. I'll reopen this issue and have a look at it later today.

**Comment by maxiheyner** [comment]
Issue tracker comment

**Comment by MateStrysewske** [comment]
Saving a case seems to fail with the following rollback output:

```
Constraint violations: 
 (1) Kind: PARAMETER
 parameter index: 0
 message: darf nicht null sein
 root bean: de.symeda.sormas.backend.person.PersonContactDetailService@4453e9a3
 property path: getByUuid.arg0
```

**Comment by FredrikSchaeferVitagroup** [comment]
I'm currently testing the climedo interface on https://ga1.sormas-oegd.de/.
When I want to register a person in climedo, at least an email adress or a phone number needs to be provided.
When I create a new case and add an email adress a contactDetail, and then click the register in climedo button, I get the following error message:
![image](https://user-images.githubusercontent.com/67001822/111970034-252c0b80-8afb-11eb-97a3-77bafd1f9082.png)
I have to set the email adress as the primary contact details, then the registration works. Is this intended behaviour?

Should we not either display another error message here? Or set the contact detail as primary, if it's the only contact information?

Please feel free to move this to another issue if you find that more appropriate.

**Comment by BarnaBartha** [comment]
@FredrikSchaeferVitagroup it is intended behaviour, a case person can have primary contact detail for each type (phone/email/other). 
Setting the contact detail as primary of it is the only one is possible, and probably makes sense. @MateStrysewske what do you think ?

**Comment by MateStrysewske** [comment]
@BarnaBartha @FredrikSchaeferVitagroup How exactly should this be done? In retrospection for all existing database entries? An alternative would be to check when registering the person in Climedo whether there's only one contact detail of the respective type that is NOT third party and just use that no matter whether it's primary or not.

Concerning the error message, I agree that it should be extended to clarify that a primary email address or phone number is required (because the error message would still pop up e.g. when a person has two phone numbers, none of which are primary).

**Comment by FredrikSchaeferVitagroup** [comment]
> @BarnaBartha @FredrikSchaeferVitagroup How exactly should this be done? In retrospection for all existing database entries? An alternative would be to check when registering the person in Climedo whether there's only one contact detail of the respective type that is NOT third party and just use that no matter whether it's primary or not.
> 
For me, both would be okay. The second option has the benefit that, for newly created persons, users don't need to define the primary contact detail if just one detail is provided.

> Concerning the error message, I agree that it should be extended to clarify that a primary email address or phone number is required (because the error message would still pop up e.g. when a person has two phone numbers, none of which are primary).

I we decide to go for the second option, there should also be no need to extend the error message.

**Comment by BarnaBartha** [comment]
@MateStrysewske i don't think we need to do this retrospectively, the migration already does this (that is moves the existing phone number or email address to a primary PersonContactDetail

**Comment by MateStrysewske** [comment]
@FredrikSchaeferVitagroup I think the error message is still required in case there are multiple contact details of the same type, with none of them being selected as primary. Because in that case, we have no way to decide which one we should transfer.

@BarnaBartha Yes, makes sense!

How are we proceeding here? @FredrikSchaeferVitagroup Do you have the capacities to build this logic for the upcoming release? I think we can associate it with this issue rather than creating a new one.

**Comment by FredrikSchaeferVitagroup** [comment]
@MateStrysewske Definitely not today. I can try to do it tomorrow, if that is still sufficient.

**Comment by MateStrysewske** [comment]
@FredrikSchaeferVitagroup Yes, absolutely :)
